File: Java/src/main/java/QuestionBank/Solution101_200/Solution144.java
package QuestionBank.Solution101_200;

import tool.TreeNode;

import java.util.ArrayList;
import java.util.List;

/**
 * @author windows
 */
public class Solution144 {

    ArrayList<Integer> list = new ArrayList<>();

    public static void main(String[] args) {
        Solution144 solution144 = new Solution144();
    }

    public List<Integer> preorderTraversal(TreeNode root) {
        if (root == null) {
            return list;
        }
        list.add(root.val);
        preorderTraversal(root.left);
        preorderTraversal(root.right);
        return list;
    }
}

File: C++/src/Solution101_200/Solution144.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution144.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
#include <functional>

using namespace std;

//vector<int> Solution144::preorderTraversal(TreeNode *root) {
//    vector<int> res;
//    dfs(root, res);
//    return res;
//}
//
//void Solution144::dfs(TreeNode *root, vector<int> &res) {
//    if (root = nullptr){
//        return;
//    }
//    res.push_back(root->val);
//    dfs(root->left, res);
//    dfs(root->right, res);
//}



vector<int> preorderTraversal(TreeNode* root) {
    vector<int> ans;
    function<void(TreeNode*)> dfs = [&](TreeNode* node){
        if (node == nullptr){
            return;
        }
        ans.push_back(node->val);
        dfs(node->left);
        dfs(node->right);
    };
    dfs(root);
    return ans;
}
